src/Compilers/Core/MSBuildTask/MapSourceRoots.cs

```diff
@@ -0,0 +1,223 @@
+            public const string NestedRoot = nameof(NestedRoot);
+            public const string ContainingRoot = nameof(ContainingRoot);
+            public const string MappedPath = nameof(MappedPath);
+            public const string SourceLinkUrl = nameof(SourceLinkUrl);
+
+            public static readonly string[] MetadataNames = new[] { SourceControl, RevisionId, NestedRoot, ContainingRoot, MappedPath, SourceLinkUrl };
+        }
+
+        private static string EndWithSlash(string path)
+            => (path[path.Length - 1] == '/') ? path : path + '/';
+
+        private static bool EndsWithDirectorySeparator(string path)
+        {
+            if (path.Length == 0)
+            {
+                return false;
+            }
+
+            char c = path[path.Length - 1];
+            return c == Path.DirectorySeparatorChar || c == Path.AltDirectorySeparatorChar;
+        }
+
+        private void MergeSourceRootMetadata(ITaskItem left, ITaskItem right)
+        {
+            foreach (string metadataName in right.MetadataNames)
+            {
+                var leftValue = left.GetMetadata(metadataName);
+                var rightValue = right.GetMetadata(metadataName);
+
+                if (!string.IsNullOrEmpty(leftValue) && !string.IsNullOrEmpty(rightValue))
+                {
+                    Log.LogErrorFromResources("MapSourceRoots.ContainsDuplicate", Names.SourceRoot, right.ItemSpec, metadataName, leftValue, rightValue);
+                }
+                else if (string.IsNullOrEmpty(leftValue) && !string.IsNullOrEmpty(rightValue))
+                {
+                    left.SetMetadata(metadataName, rightValue);
+                }
+            }
+        }
+
+        public override bool Execute()
+        {
+            // Merge metadata of SourceRoot items with the same identity.
+            var mappedSourceRoots = new List<ITaskItem>();
+            var rootByItemSpec = new Dictionary<string, ITaskItem>();
+            foreach (var sourceRoot in SourceRoots)
+            {
+                // The SourceRoot is required to have a trailing directory separator.
+                // We do not append one implicitly as we do not know which separator to append on Windows.
+                // The usage of SourceRoot might be sensitive to what kind of separator is used (e.g. in SourceLink where it needs
+                // to match the corresponding separators used in paths given to the compiler).
+                if (!EndsWithDirectorySeparator(sourceRoot.ItemSpec))
+                {
+                    Log.LogErrorFromResources("MapSourceRoots.PathMustEndWithSlashOrBackslash", Names.SourceRoot, sourceRoot.ItemSpec);
+                }
+
+                if (rootByItemSpec.TryGetValue(sourceRoot.ItemSpec, out var existingRoot))
+                {
+                    ReportConflictingWellKnownMetadata(existingRoot, sourceRoot);
+                    sourceRoot.CopyMetadataTo(existingRoot);
```

Comment:
Is it possible there would be conflict of metadata, or do we expect that different sources would be consistent?

Reply:
We do expect them to be consistent, but in theory there may be conflict. One can always specify these values manually in proj file. We warn if we find inconsistency.